### DISCUSSION TITLE
When to Open a Session Reasoning

### DISCUSSION THREAD

**Discussion by mark-barrett** [sqlalchemy/sqlalchemy#9160] [Usage Questions] [answered]
Over the last few weeks we have had issues that we have since resolved with answers on the discussions here on the SQLAlchemy GitHub.

A big one for us was that we had an endpoint that on the outset was quite simple, it would create an instance of a model. However the implementation of the save required a long list of Selects to get data, updates, inserts, append to relationships etc.

For this, we used what FastAPI has in it's examples:

```python
@app.post("/resource", response_model=ResourceSchema)
def create_resource(resource: ResourceSchema, db: Session = Depends(get_db)):
    # logic to handle the create
```

Creating and yielding a session at the start of the request and using the same one throughout the entire request.

From reading the SQLAlchemy docs, I can see that this is also recommended here:

https://docs.sqlalchemy.org/en/14/orm/session_basics.html#when-do-i-construct-a-session-when-do-i-commit-it-and-when-do-i-close-it

where it says

_the basic pattern is create a Session at the start of a web request, call the Session.commit() method at the end of web requests that do POST, PUT, or DELETE_

What we found was because we did a huge list of different queries, when we added functionality whereby we wanted to catch a Unique Constraint then roll it back, it caused a whole host of issues. 

I'm not sure if perhaps I'm misunderstanding it, but from what I saw, the rollback rolled everything back in that session and unless I expunged the objects, they were no longer available.

I understand this might be intended as the request is a single atomic thing, however the particular thing we were rolling back was a simple "create, if fails, rollback and update instead". I've since changed this to a `insert.on_conflict.do_update`.

My query question is, why is it advised to have a single session for the entire request? Why is something like this not advised:

```python
@app.post("/resource", response_model=ResourceSchema)
def create_resource(resource: ResourceSchema):
    async with db_session() as db:
        db.add(self)
        await db.commit()

    async with db_session() as db:
        db.add(self)
        await db.commit()
```

I don't mean doing it exactly like that, but I mean, letting the session be opened and closed multiple times in functions in the endpoint.

Does this solve my initial rollback issue? For example, if there was an error in the second context manager and I called rollback, it'd only rollback what was done in that particular session right?

I apologise in advance if it's an already answered or dumb question.

**Comment by zzzeek** [comment, marked as answer]
there's no "rule" and it depends on what you want the application to actually do when things go wrong.     The question here seems to be, "what is the purpose of atomicity".   Certainly if you have some kind of operation that does several things, and you *want* each element to be permanently committed even if the Nth element fails, then you would use either autocommit, or smaller transaction blocks.

In the typical web service use case, a POST or PUT is doing something small.  It might involve one row, or a few.   It is generally highly desirable that if part of the operation fails, the whole thing should fail, so that the entire operation can be attempted again.  Otherwise, only half the data that the operation planned to add to the database would be stored in persistent storage and the other half would be non-present.   It's typical that an application might not know how to deal with partially committed data on the next run.

in the case you have, where you are doing an upsert, that's not actually an operation you are assuming has "failed", it's one where you need to perform one of two different operations depending on database state.    The "on conflict" you are doing is absolutely the appropriate way to handle that.  There's another way to handle it as well, which is to use a SAVEPOINT.    That's the general way to "try" an operation that might fail, but you dont want the transaction to fail.   This pattern is documented at https://docs.sqlalchemy.org/en/20/orm/session_transaction.html#session-begin-nested .
